File: sdk/flutter/test/realtime_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:loomup/loomup.dart';
import 'package:test/test.dart';

import 'mocks.dart';

Map<String, dynamic>? _tryParse(String text) {
  try {
    final decoded = jsonDecode(text);
    if (decoded is Map) return Map<String, dynamic>.from(decoded);
  } catch (_) {}
  return null;
}

/// Factory that auto-acks subscribe frames with `subscribed`.
WebSocketFactory autoAckFactory({
  void Function(MockWebSocket ws)? onCreated,
  bool errorAck = false,
}) {
  return () {
    final ws = MockWebSocket();
    onCreated?.call(ws);
    ws.onSend = (text) {
      final obj = _tryParse(text);
      if (obj != null && obj['type'] == 'subscribe') {
        final rid = obj['requestId'];
        scheduleMicrotask(() {
          if (errorAck) {
            ws.simulateMessage(
              jsonEncode({
                'type': 'error',
                'requestId': rid,
                'code': 'SUBSCRIBE_ERROR',
                'message': 'forbidden',
              }),
            );
          } else {
            ws.simulateMessage(
              jsonEncode({
                'type': 'subscribed',
                'requestId': rid,
                'table': obj['table'],
              }),
            );
          }
        });
      }
    };
    return ws;
  };
}

void main() {
  test('injected WebSocket constructed on subscribe', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    expect(box.socket, isNull);
    final unsub = c.from('todos').subscribe((_) {});
    await Future<void>.delayed(const Duration(milliseconds: 20));
    expect(box.socket, isNotNull);
    expect(box.socket!.connectCount, 1);
    unsub();
    c.closeRealtime();
  });

  test('control error frames surface with code', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    final controls = <ControlEvent>[];
    final off = c.onControl(controls.add);
    final unsub = c.from('todos').subscribe((_) {});
    await Future<void>.delayed(const Duration(milliseconds: 20));
    box.socket?.simulateMessage(
      '{"type":"error","code":"AUTH_ERROR","message":"invalid or expired token"}',
    );
    box.socket?.simulateMessage(
      '{"type":"error","code":"SUBSCRIBE_ERROR","table":"todos","message":"subscribe forbidden"}',
    );
    await Future<void>.delayed(const Duration(milliseconds: 10));
    expect(
      controls.any((e) => e.type == 'error' && e.code == 'AUTH_ERROR'),
      isTrue,
    );
    expect(
      controls.any((e) => e.type == 'error' && e.code == 'SUBSCRIBE_ERROR'),
      isTrue,
    );
    off();
    unsub();
    c.closeRealtime();
  });

  test('setToken reauths and resubscribes', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      token: 'old-token',
      webSocketFactory: box.factory(),
    );
    final unsub = c.from('todos').subscribe((_) {}, rowId: '1');
    await Future<void>.delayed(const Duration(milliseconds: 20));
    box.socket?.clearSent();
    c.setToken('new-token');
    await Future<void>.delayed(const Duration(milliseconds: 10));
    final frames = box.socket?.parsedSent() ?? [];
    expect(
      frames.any((f) => f['type'] == 'auth' && f['token'] == 'new-token'),
      isTrue,
      reason: '$frames',
    );
    expect(
      frames.any((f) => f['type'] == 'subscribe' && f['token'] == 'new-token'),
      isTrue,
      reason: '$frames',
    );
    unsub();
    c.closeRealtime();
  });

  test('row unsub sends id', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    final u1 = c.from('todos').subscribe((_) {}, rowId: '1');
    final u2 = c.from('todos').subscribe((_) {}, rowId: '2');
    await Future<void>.delayed(const Duration(milliseconds: 20));
    u1();
    await Future<void>.delayed(const Duration(milliseconds: 10));
    final unsubs = (box.socket?.parsedSent() ?? [])
        .where((f) => f['type'] == 'unsubscribe')
        .toList();
    expect(unsubs.any((f) => f['id'] == '1'), isTrue, reason: '$unsubs');
    u2();
    c.closeRealtime();
  });

  test('subscribe with hash in rowId', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      token: 'tok1',
      webSocketFactory: box.factory(),
    );
    const rowId = 'prefix#with#hashes';
    final unsub = c.from('items').subscribe((_) {}, rowId: rowId);
    await Future<void>.delayed(const Duration(milliseconds: 20));
    final subs = (box.socket?.parsedSent() ?? [])
        .where((f) => f['type'] == 'subscribe')
        .toList();
    expect(
      subs.any((f) => f['id'] == rowId && f['table'] == 'items'),
      isTrue,
      reason: '$subs',
    );
    unsub();
    c.closeRealtime();
  });

  test('change events fan out to table and row handlers', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    final tableEvents = <ChangeEvent>[];
    final rowEvents = <ChangeEvent>[];
    final u1 = c.from('todos').subscribe(tableEvents.add);
    final u2 = c.from('todos').subscribe(rowEvents.add, rowId: '9');
    await Future<void>.delayed(const Duration(milliseconds: 20));
    box.socket?.simulateMessage(
      '{"type":"change","table":"todos","op":"INSERT","id":"9","data":{"id":9,"title":"x"},"ts":100}',
    );
    await Future<void>.delayed(const Duration(milliseconds: 10));
    expect(tableEvents.length, 1);
    expect(rowEvents.length, 1);
    expect(tableEvents.first.op, 'INSERT');
    expect(rowEvents.first.id, '9');
    u1();
    u2();
    c.closeRealtime();
  });

  test('subscribeReady resolves on subscribed ack', () async {
    MockWebSocket? last;
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: autoAckFactory(onCreated: (ws) => last = ws),
    );

    final unsub = await c.from('todos').subscribeReady((_) {});
    expect(unsub, isNotNull);
    expect(last, isNotNull);
    unsub();
    c.closeRealtime();
  });

  test('subscribeReady rejects on error frame', () async {
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: autoAckFactory(errorAck: true),
    );

    try {
      await c.from('todos').subscribeReady((_) {});
      fail('expected throw');
    } on LoomupException catch (e) {
      expect(e.message, 'forbidden');
      expect(e.code, 'SUBSCRIBE_ERROR');
    }
    c.closeRealtime();
  });

  test('refresh applyTokens sends auth and resubscribe', () async {
    final http = MockHttp();
    http.handler = (method, url, auth, body) async {
      if (url.endsWith('/auth/refresh')) {
        return (
          body: jsonBytes({
            'data': {
              'access_token': 'rotated-access',
              'refresh_token': 'rotated-refresh',
              'token_type': 'Bearer',
              'expires_in': 900,
            },
          }),
          status: 200,
        );
      }
      return (body: 'nope'.codeUnits, status: 500);
    };
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://example.test',
      token: 'old-access',
      refreshToken: 'r1',
      http: http,
      webSocketFactory: box.factory(),
    );
    final unsub = c.from('todos').subscribe((_) {}, rowId: '42');
    await Future<void>.delayed(const Duration(milliseconds: 20));
    if (box.socket?.isOpen != true) {
      box.socket?.simulateOpen();
    }
    box.socket?.clearSent();
    await c.refresh();
    expect(c.accessToken, 'rotated-access');
    await Future<void>.delayed(const Duration(milliseconds: 10));
    final frames = box.socket?.parsedSent() ?? [];
    expect(
      frames.any(
        (f) => f['type'] == 'auth' && f['token'] == 'rotated-access',
      ),
      isTrue,
      reason: '$frames',
    );
    expect(
      frames.any(
        (f) =>
            f['type'] == 'subscribe' &&
            f['token'] == 'rotated-access' &&
            f['id'] == '42',
      ),
      isTrue,
      reason: '$frames',
    );
    unsub();
    c.closeRealtime();
  });

  test('unexpected close schedules reconnect', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    final unsub = c.from('todos').subscribe((_) {});
    await Future<void>.delayed(const Duration(milliseconds: 20));
    expect(box.totalConnectCount, 1);
    box.socket!.simulateClose();
    // First reconnect: full jitter in [50ms, 1000ms] — wait past the cap.
    await Future<void>.delayed(const Duration(milliseconds: 1200));
    expect(box.totalConnectCount, greaterThanOrEqualTo(2));
    unsub();
    c.closeRealtime();
  });

  test('closeRealtime stops reconnect and fails pending acks', () async {
    final box = MockWebSocketBox()..autoOpen = false;
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    final ready = c.from('todos').subscribeReady((_) {}, timeoutMs: 2000);
    await Future<void>.delayed(const Duration(milliseconds: 30));
    c.closeRealtime();
    try {
      await ready;
      fail('expected throw');
    } on LoomupException catch (e) {
      expect(e.code, 'realtime_closed');
    }
  });

  test('RESYNC after reconnect for row subscription', () async {
    final http = MockHttp();
    http.handler = (method, url, auth, body) async {
      if (url.contains('/api/todos/7')) {
        return (
          body: jsonBytes({
            'data': {'id': 7, 'title': 'catch-up'},
          }),
          status: 200,
        );
      }
      return (body: 'nope'.codeUnits, status: 404);
    };
    final box = MockWebSocketBox();
    final events = <ChangeEvent>[];
    final c = createClient(
      url: 'http://localhost:3000',
      http: http,
      webSocketFactory: box.factory(),
    );
    final unsub = c.from('todos').subscribe(events.add, rowId: '7');
    await Future<void>.delayed(const Duration(milliseconds: 20));
    // Simulate disconnect + reconnect (second open triggers RESYNC).
    // First reconnect delay is full-jitter up to 1s.
    box.socket!.simulateClose();
    await Future<void>.delayed(const Duration(milliseconds: 1300));
    expect(events.any((e) => e.op == 'RESYNC' && e.id == '7'), isTrue);
    unsub();
    c.closeRealtime();
  });

  test('changes stream delivers and unsubscribes on cancel', () async {
    final box = MockWebSocketBox();
    final c = createClient(
      url: 'http://localhost:3000',
      webSocketFactory: box.factory(),
    );
    final events = <ChangeEvent>[];
    final sub = c.from('todos').changes().listen(events.add);
    await Future<void>.delayed(const Duration(milliseconds: 20));
    box.socket?.simulateMessage(
      '{"type":"change","table":"todos","op":"UPDATE","id":"1","data":{"id":1},"ts":1}',
    );
    await Future<void>.delayed(const Duration(milliseconds: 10));
    expect(events.length, 1);
    await sub.cancel();
    final unsubs = (box.socket?.parsedSent() ?? [])
        .where((f) => f['type'] == 'unsubscribe')
        .toList();
    expect(unsubs, isNotEmpty);
    c.closeRealtime();
  });
}

File: kotlin/src/test/kotlin/com/loomup/client/RealtimeTest.kt
package com.loomup.client

import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class RealtimeTest {
    @Test
    fun injectedWebSocketConstructedOnSubscribe() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://localhost:3000",
            webSocketFactory = box.factory(),
        )
        // No socket until subscribe
        val unsub = c.from("todos").subscribe { }
        delay(50)
        assertNotNull(box.socket)
        assertEquals(1, box.socket!!.connectCountValue)
        unsub()
        c.closeRealtime()
    }

    @Test
    fun controlErrorFramesSurfaceWithCode() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://localhost:3000",
            webSocketFactory = box.factory(),
        )
        val controls = mutableListOf<ControlEvent>()
        val off = c.onControl { controls.add(it) }
        val unsub = c.from("todos").subscribe { }
        delay(50)
        box.socket?.simulateMessage(
            """{"type":"error","code":"AUTH_ERROR","message":"invalid or expired token"}""",
        )
        box.socket?.simulateMessage(
            """{"type":"error","code":"SUBSCRIBE_ERROR","table":"todos","message":"subscribe forbidden"}""",
        )
        delay(20)
        assertTrue(controls.any { it.type == "error" && it.code == "AUTH_ERROR" })
        assertTrue(controls.any { it.type == "error" && it.code == "SUBSCRIBE_ERROR" })
        off()
        unsub()
        c.closeRealtime()
    }

    @Test
    fun setTokenReauthsAndResubscribes() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://localhost:3000",
            token = "old-token",
            webSocketFactory = box.factory(),
        )
        val unsub = c.from("todos").subscribe(rowId = "1") { }
        delay(50)
        box.socket?.clearSent()
        c.setToken("new-token")
        delay(20)
        val frames = box.socket?.parsedSent().orEmpty()
        assertTrue(
            frames.any {
                it["type"]?.jsonPrimitive?.contentOrNull == "auth" &&
                    it["token"]?.jsonPrimitive?.contentOrNull == "new-token"
            },
            frames.toString(),
        )
        assertTrue(
            frames.any {
                it["type"]?.jsonPrimitive?.contentOrNull == "subscribe" &&
                    it["token"]?.jsonPrimitive?.contentOrNull == "new-token"
            },
            frames.toString(),
        )
        unsub()
        c.closeRealtime()
    }

    @Test
    fun rowUnsubSendsId() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://localhost:3000",
            webSocketFactory = box.factory(),
        )
        val u1 = c.from("todos").subscribe(rowId = "1") { }
        val u2 = c.from("todos").subscribe(rowId = "2") { }
        delay(40)
        u1()
        delay(20)
        val unsubs = (box.socket?.parsedSent().orEmpty())
            .filter { it["type"]?.jsonPrimitive?.contentOrNull == "unsubscribe" }
        assertTrue(
            unsubs.any { it["id"]?.jsonPrimitive?.contentOrNull == "1" },
            unsubs.toString(),
        )
        u2()
        c.closeRealtime()
    }

    @Test
    fun subscribeWithHashInRowId() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://localhost:3000",
            token = "tok1",
            webSocketFactory = box.factory(),
        )
        val rowId = "prefix#with#hashes"
        val unsub = c.from("items").subscribe(rowId = rowId) { }
        delay(50)
        val subs = (box.socket?.parsedSent().orEmpty())
            .filter { it["type"]?.jsonPrimitive?.contentOrNull == "subscribe" }
        assertTrue(
            subs.any {
                it["id"]?.jsonPrimitive?.contentOrNull == rowId &&
                    it["table"]?.jsonPrimitive?.contentOrNull == "items"
            },
            subs.toString(),
        )
        unsub()
        c.closeRealtime()
    }

    @Test
    fun refreshApplyTokensSendsAuthAndResubscribe() = runBlocking {
        val http = MockHttp()
        http.handler = { _, url, _, _ ->
            if (url.endsWith("/auth/refresh")) {
                val body = """
                    {
                      "data": {
                        "access_token": "rotated-access",
                        "refresh_token": "rotated-refresh",
                        "token_type": "Bearer",
                        "expires_in": 900
                      }
                    }
                """.trimIndent()
                jsonBytes(body) to 200
            } else {
                jsonBytes("nope") to 500
            }
        }
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://example.test",
            token = "old-access",
            refreshToken = "r1",
            http = http,
            webSocketFactory = box.factory(),
        )
        val unsub = c.from("todos").subscribe(rowId = "42") { }
        delay(50)
        if (box.socket?.isOpen != true) {
            box.socket?.simulateOpen()
        }
        box.socket?.clearSent()
        c.refresh()
        assertEquals("rotated-access", c.accessToken)
        delay(30)
        val frames = box.socket?.parsedSent().orEmpty()
        assertTrue(
            frames.any {
                it["type"]?.jsonPrimitive?.contentOrNull == "auth" &&
                    it["token"]?.jsonPrimitive?.contentOrNull == "rotated-access"
            },
            frames.toString(),
        )
        assertTrue(
            frames.any {
                it["type"]?.jsonPrimitive?.contentOrNull == "subscribe" &&
                    it["token"]?.jsonPrimitive?.contentOrNull == "rotated-access" &&
                    it["table"]?.jsonPrimitive?.contentOrNull == "todos" &&
                    it["id"]?.jsonPrimitive?.contentOrNull == "42"
            },
            frames.toString(),
        )
        unsub()
        c.closeRealtime()
    }

    @Test
    fun reconnectResyncDeliversResyncEvents() = runBlocking {
        val http = MockHttp()
        http.handler = { _, url, _, _ ->
            if (url.contains("/api/todos/7")) {
                jsonBytes("""{"data":{"id":7,"title":"after-outage"}}""") to 200
            } else {
                jsonBytes("nope") to 404
            }
        }
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://example.test",
            token = "t",
            http = http,
            webSocketFactory = box.factory(),
        )
        val events = mutableListOf<ChangeEvent>()
        val unsub = c.from("todos").subscribe(rowId = "7") { events.add(it) }
        delay(50)
        assertEquals(0, events.count { it.op == "RESYNC" })

        // Drop + reconnect
        box.socket?.simulateClose()
        // Wait for reconnect timer (min ~50ms, jitter up to 1s first attempt)
        delay(1300)

        val resyncs = events.filter { it.op == "RESYNC" }
        assertTrue(resyncs.isNotEmpty(), "events=${events.map { it.op }}")
        val first = resyncs.first()
        assertEquals("7", first.id)
        assertEquals("after-outage", first.data?.get("title")?.stringValue)
        unsub()
        c.closeRealtime()
    }

    @Test
    fun subscribeReadyAwaitsSubscribedAck() = runBlocking {
        val box = MockWebSocketBox()
        val factory: WebSocketFactory = {
            val ws = MockWebSocket()
            ws.onSend = { text ->
                try {
                    val json = kotlinx.serialization.json.Json.parseToJsonElement(text).jsonObjectOrNull()
                    if (json?.get("type")?.jsonPrimitive?.contentOrNull == "subscribe") {
                        val rid = json["requestId"]?.jsonPrimitive?.contentOrNull
                        if (rid != null) {
                            val ack =
                                """{"type":"subscribed","requestId":"$rid","table":"todos","channel":"todos"}"""
                            Thread {
                                Thread.sleep(5)
                                ws.simulateMessage(ack)
                            }.start()
                        }
                    }
                } catch (_: Exception) {
                }
            }
            box.note(ws)
            ws
        }
        val c = createClient(
            url = "http://example.test",
            webSocketFactory = factory,
        )
        val unsub = c.from("todos").subscribeReady { }
        unsub()
        c.closeRealtime()
    }

    @Test
    fun subscribeReadyTimeoutCleansUp() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://example.test",
            webSocketFactory = box.factory(),
        )
        val e = assertFailsWith<LoomupError> {
            c.from("todos").subscribeReady(timeoutMs = 80) { }
        }
        assertTrue(
            e.message!!.contains("timeout") || e.code == "subscribe_timeout",
            e.message,
        )
        c.closeRealtime()
    }

    @Test
    fun subscribeReadyRejectsOnErrorFrame() = runBlocking {
        val box = MockWebSocketBox()
        val factory: WebSocketFactory = {
            val ws = MockWebSocket()
            ws.onSend = { text ->
                try {
                    val json = kotlinx.serialization.json.Json.parseToJsonElement(text).jsonObjectOrNull()
                    if (json?.get("type")?.jsonPrimitive?.contentOrNull == "subscribe") {
                        val rid = json["requestId"]?.jsonPrimitive?.contentOrNull
                        if (rid != null) {
                            val err =
                                """{"type":"error","code":"SUBSCRIBE_ERROR","requestId":"$rid","message":"table not exposed or realtime disabled"}"""
                            Thread {
                                Thread.sleep(5)
                                ws.simulateMessage(err)
                            }.start()
                        }
                    }
                } catch (_: Exception) {
                }
            }
            box.note(ws)
            ws
        }
        val c = createClient(
            url = "http://example.test",
            webSocketFactory = factory,
        )
        val e = assertFailsWith<LoomupError> {
            c.from("todos").subscribeReady(timeoutMs = 2000) { }
        }
        assertTrue(
            e.message!!.contains("table not exposed") ||
                e.code == "SUBSCRIBE_ERROR" ||
                e.message!!.contains("subscribe"),
            e.message,
        )
        c.closeRealtime()
    }

    @Test
    fun changeEventFanout() = runBlocking {
        val box = MockWebSocketBox()
        val c = createClient(
            url = "http://example.test",
            webSocketFactory = box.factory(),
        )
        val tableEvents = mutableListOf<ChangeEvent>()
        val rowEvents = mutableListOf<ChangeEvent>()
        val u1 = c.from("todos").subscribe { tableEvents.add(it) }
        val u2 = c.from("todos").subscribe(rowId = "9") { rowEvents.add(it) }
        delay(40)
        box.socket?.simulateMessage(
            """{"type":"change","table":"todos","op":"INSERT","id":"9","data":{"id":9,"title":"x"},"ts":100}""",
        )
        delay(20)
        assertEquals(1, tableEvents.size)
        assertEquals(1, rowEvents.size)
        assertEquals("INSERT", tableEvents.first().op)
        u1()
        u2()
        c.closeRealtime()
    }
}

private fun kotlinx.serialization.json.JsonElement.jsonObjectOrNull():
    kotlinx.serialization.json.JsonObject? =
    this as? kotlinx.serialization.json.JsonObject
